File: routa-core/src/main/kotlin/com/phodal/routa/core/runner/RoutaOrchestrator.kt
package com.phodal.routa.core.runner

import com.phodal.routa.core.RoutaSystem
import com.phodal.routa.core.pipeline.OrchestrationPipeline
import com.phodal.routa.core.pipeline.PipelineContext
import com.phodal.routa.core.pipeline.PipelineEventBridge
import com.phodal.routa.core.provider.AgentProvider
import com.phodal.routa.core.provider.AgentRunnerAdapter
import com.phodal.routa.core.provider.StreamChunk
import com.phodal.routa.core.report.ReportParser
import com.phodal.routa.core.report.TextPatternReportParser
import kotlinx.coroutines.Job
import kotlin.coroutines.coroutineContext

/**
 * The multi-agent orchestration entry point.
 *
 * Delegates to a composable [OrchestrationPipeline] that executes
 * independent, testable stages in sequence. The default pipeline
 * implements the full ROUTA → CRAFTER → GATE workflow:
 *
 * ```
 * User Request
 *   → [PlanningStage] ROUTA plans (@@@task blocks)
 *     → [TaskRegistrationStage] Parse and register tasks
 *       → [CrafterExecutionStage] CRAFTER agents execute tasks
 *         → [GateVerificationStage] GATE verifies work
 *           → APPROVED: done
 *           → NOT APPROVED: repeat from CrafterExecution
 * ```
 *
 * ## Customizing the Pipeline
 *
 * ```kotlin
 * // Use the default pipeline
 * val orchestrator = RoutaOrchestrator(system, provider, "workspace")
 *
 * // Use a custom pipeline (e.g., skip verification)
 * val orchestrator = RoutaOrchestrator(
 *     system, provider, "workspace",
 *     pipeline = OrchestrationPipeline.withoutVerification(),
 * )
 *
 * // Add custom stages
 * val orchestrator = RoutaOrchestrator(
 *     system, provider, "workspace",
 *     pipeline = OrchestrationPipeline.default()
 *         .withStageAt(3, CodeReviewStage()),
 * )
 * ```
 *
 * ## Backward Compatibility
 *
 * This class maintains the same public API as before. The [AgentRunner]
 * parameter is accepted but must be an [AgentProvider] instance.
 * All streaming, phase callbacks, and result types are unchanged.
 *
 * @param routa The Routa multi-agent system.
 * @param runner The agent execution provider (must implement [AgentProvider]).
 * @param workspaceId The workspace to operate in.
 * @param maxWaves Maximum pipeline iterations (default: 3). Overridden by pipeline config if provided.
 * @param parallelCrafters Whether CRAFTERs execute in parallel within a wave.
 * @param onPhaseChange Callback for orchestration phase changes.
 * @param onStreamChunk Callback for agent streaming chunks.
 * @param pipeline Custom pipeline. If null, uses [OrchestrationPipeline.default].
 * @param reportParser Custom report parser. If null, uses [TextPatternReportParser].
 */
class RoutaOrchestrator(
    private val routa: RoutaSystem,
    private val runner: AgentRunner,
    private val workspaceId: String,
    private val maxWaves: Int = 3,
    private val parallelCrafters: Boolean = false,
    private val onPhaseChange: (suspend (OrchestratorPhase) -> Unit)? = null,
    private val onStreamChunk: ((agentId: String, chunk: StreamChunk) -> Unit)? = null,
    private val pipeline: OrchestrationPipeline? = null,
    private val reportParser: ReportParser? = null,
) {

    // Wrap legacy AgentRunner in an adapter if needed
    private val provider: AgentProvider = when (runner) {
        is AgentProvider -> runner
        else -> AgentRunnerAdapter(runner)
    }

    /**
     * The resolved pipeline — either the one passed in, or the default.
     */
    val resolvedPipeline: OrchestrationPipeline =
        pipeline ?: OrchestrationPipeline.default(maxWaves)

    /**
     * The resolved report parser — either the one passed in, or the default.
     */
    val resolvedReportParser: ReportParser =
        reportParser ?: TextPatternReportParser()

    /**
     * The pipeline event bridge — exposes pipeline lifecycle events for
     * the collaboration plane (UI, metrics, recovery handlers).
     *
     * ```kotlin
     * orchestrator.eventBridge.subscribeTo<PipelineEvent.StageCompleted>()
     *     .collect { event -> logger.info("Stage ${event.stageName} took ${event.durationMs}ms") }
     * ```
     */
    val eventBridge: PipelineEventBridge
        get() = resolvedPipeline.eventBridge

    /**
     * Execute the full multi-agent orchestration flow.
     *
     * Creates a [PipelineContext] and delegates execution to the [OrchestrationPipeline].
     * The current coroutine's [Job] is passed to the context for cancellation propagation —
     * when the caller's scope is cancelled, the pipeline exits cooperatively.
     *
     * @param userRequest The user's requirement/task description.
     * @return The orchestration result.
     */
    suspend fun execute(userRequest: String): OrchestratorResult {
        val context = PipelineContext(
            system = routa,
            provider = provider,
            workspaceId = workspaceId,
            userRequest = userRequest,
            reportParser = resolvedReportParser,
            parallelCrafters = parallelCrafters,
            onPhaseChange = onPhaseChange,
            onStreamChunk = onStreamChunk,
            parentJob = coroutineContext[Job],
        )

        return resolvedPipeline.execute(context)
    }
}

// ── Orchestrator phases (for UI/CLI callbacks) ──────────────────────────

/**
 * Phases emitted during orchestration for UI/CLI observation.
 *
 * These phases are emitted by pipeline stages via [PipelineContext.emitPhase]
 * and forwarded to the caller via the [RoutaOrchestrator.onPhaseChange] callback.
 */
sealed class OrchestratorPhase {
    /** Pipeline is initializing. */
    data object Initializing : OrchestratorPhase()

    /** ROUTA is planning tasks. */
    data object Planning : OrchestratorPhase()

    /** ROUTA's plan output is ready. */
    data class PlanReady(val planOutput: String) : OrchestratorPhase()

    /** Tasks have been parsed and registered. */
    data class TasksRegistered(val count: Int) : OrchestratorPhase()

    /** A new wave of CRAFTER execution is starting. */
    data class WaveStarting(val wave: Int) : OrchestratorPhase()

    /** A specific CRAFTER is running. */
    data class CrafterRunning(val crafterId: String, val taskId: String) : OrchestratorPhase()

    /** A specific CRAFTER completed. */
    data class CrafterCompleted(val crafterId: String, val taskId: String) : OrchestratorPhase()

    /** GATE verification is starting. */
    data class VerificationStarting(val wave: Int) : OrchestratorPhase()

    /** GATE verification completed. */
    data class VerificationCompleted(val gateId: String, val output: String) : OrchestratorPhase()

    /** GATE found issues, tasks need fixes. */
    data class NeedsFix(val wave: Int) : OrchestratorPhase()

    /** All tasks completed successfully. */
    data object Completed : OrchestratorPhase()

    /** Maximum iterations reached without full approval. */
    data class MaxWavesReached(val waves: Int) : OrchestratorPhase()
}

// ── Orchestrator results ────────────────────────────────────────────────

/**
 * The final outcome of an orchestration run.
 */
sealed class OrchestratorResult {
    /** All tasks completed and verified successfully. */
    data class Success(
        val taskSummaries: List<com.phodal.routa.core.coordinator.TaskSummary>,
    ) : OrchestratorResult()

    /** ROUTA's plan contained no @@@task blocks. */
    data class NoTasks(val planOutput: String) : OrchestratorResult()

    /** Maximum iterations exhausted without full approval. */
    data class MaxWavesReached(
        val waves: Int,
        val taskSummaries: List<com.phodal.routa.core.coordinator.TaskSummary>,
    ) : OrchestratorResult()

    /** Pipeline failed with an error. */
    data class Failed(val error: String) : OrchestratorResult()
}

File: routa-js/src/core/orchestrator/routa-orchestrator.ts
/**
 * RoutaOrchestrator - port of routa-core RoutaOrchestrator.kt
 *
 * The multi-agent orchestration entry point.
 */

import { AgentProvider, StreamChunk } from "../provider/agent-provider";
import { AgentExecutionContext } from "../coordinator/routa-coordinator";
import {
  PipelineContext,
  createPipelineContext,
  OrchestratorPhase,
} from "../pipeline/pipeline-context";
import { OrchestrationPipeline } from "./orchestration-pipeline";
import { OrchestratorResult } from "../pipeline/stage-result";

/**
 * Configuration for the orchestrator.
 */
export interface OrchestratorConfig {
  /** The agent execution context (stores, tools, event bus) */
  context: AgentExecutionContext;

  /** The agent execution provider */
  provider: AgentProvider;

  /** The workspace identifier */
  workspaceId: string;

  /** Maximum pipeline iterations (default: 3) */
  maxWaves?: number;

  /** Whether CRAFTERs execute in parallel within a wave */
  parallelCrafters?: boolean;

  /** Callback for orchestration phase changes */
  onPhaseChange?: (phase: OrchestratorPhase) => void | Promise<void>;

  /** Callback for agent streaming chunks */
  onStreamChunk?: (agentId: string, chunk: StreamChunk) => void;

  /** Custom pipeline (if null, uses default) */
  pipeline?: OrchestrationPipeline;
}

/**
 * The multi-agent orchestration entry point.
 *
 * Delegates to a composable OrchestrationPipeline that executes
 * independent, testable stages in sequence. The default pipeline
 * implements the full ROUTA → CRAFTER → GATE workflow:
 *
 * ```
 * User Request
 *   → [PlanningStage] ROUTA plans (@@@task blocks)
 *     → [TaskRegistrationStage] Parse and register tasks
 *       → [CrafterExecutionStage] CRAFTER agents execute tasks
 *         → [GateVerificationStage] GATE verifies work
 *           → APPROVED: done
 *           → NOT APPROVED: repeat from CrafterExecution
 * ```
 */
export class RoutaOrchestrator {
  private context: AgentExecutionContext;
  private provider: AgentProvider;
  private workspaceId: string;
  private maxWaves: number;
  private parallelCrafters: boolean;
  private onPhaseChange?: (phase: OrchestratorPhase) => void | Promise<void>;
  private onStreamChunk?: (agentId: string, chunk: StreamChunk) => void;
  private pipeline: OrchestrationPipeline;

  constructor(config: OrchestratorConfig) {
    this.context = config.context;
    this.provider = config.provider;
    this.workspaceId = config.workspaceId;
    this.maxWaves = config.maxWaves ?? 3;
    this.parallelCrafters = config.parallelCrafters ?? false;
    this.onPhaseChange = config.onPhaseChange;
    this.onStreamChunk = config.onStreamChunk;
    this.pipeline =
      config.pipeline ?? OrchestrationPipeline.default(this.maxWaves);
  }

  /**
   * Execute the full multi-agent orchestration flow.
   *
   * Creates a PipelineContext and delegates execution to the OrchestrationPipeline.
   *
   * @param userRequest The user's requirement/task description
   * @returns The orchestration result
   */
  async execute(userRequest: string): Promise<OrchestratorResult> {
    const pipelineContext = createPipelineContext({
      context: this.context,
      provider: this.provider,
      workspaceId: this.workspaceId,
      userRequest,
      parallelCrafters: this.parallelCrafters,
      onPhaseChange: this.onPhaseChange,
      onStreamChunk: this.onStreamChunk,
    });

    return this.pipeline.execute(pipelineContext);
  }
}

